src/peer_digest.cc

```diff
@@ -152,19 +154,23 @@ peerDigestDestroy(PeerDigest * pd)
 
     /*
      * DPW 2007-04-12
-     * We locked the peer in peerDigestInit(), this is
+     * We locked the peer in PeerDigest constructor, this is
      * where we unlock it.  If the peer is still valid,
      * tell it that the digest is gone.
      */
-    if (cbdataReferenceValidDone(peerTmp, &p))
-        peerNoteDigestGone((CachePeer *)p);
-
-    delete pd->cd;
-    pd->host.clean();
+    if (cbdataReferenceValidDone(peerTmp, &p)) {
+        cbdataReferenceDone(((CachePeer *)p)->digest);
```

Comment:
If you have to rewire this code, please use a C++ cast.

Reply:
done.